### PR TITLE
feat: dynamic imports support in static bundle

### PR DESCRIPTION
**Pull request by zamotany**
https://github.com/callstack/haul/projects/2#card-21106089

### CODE REVIEW

File: packages/haul-core/src/webpack/ReactNativeTarget.ts [hunk 1]
```diff
@@ -10,18 +10,22 @@ import ReactNativeEnvPlugin from './plugins/ReactNativeEnvPlugin';
...
  */
-export default function ReactNativeTarget(compiler: webpack.Compiler) {
-  // Load modules using require, since we don't really care about types here.
-  const WebWorkerTemplatePlugin = require('webpack/lib/webworker/WebWorkerTemplatePlugin');
-  const NodeSourcePlugin = require('webpack/lib/node/NodeSourcePlugin');
-  const FunctionModulePlugin = require('webpack/lib/FunctionModulePlugin');
-  const LoaderTargetPlugin = require('webpack/lib/LoaderTargetPlugin');
+export default function ReactNativeTarget({
+  bundle,
+}: { bundle?: boolean } = {}) {
```
Comment (thymikee) [inline packages/haul-core/src/webpack/ReactNativeTarget.ts]: You'd avoid casting to boolean later by providing a default false value

File: packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts [hunk 2]
```diff
@@ -13,26 +15,116 @@ function asyncEval(url) {
...
+          asyncChunks,
+        } = this.getFilenamesFromChunks(compilation.chunks);
+
+        const sourceMappingRegex = /\/\/# sourceMappingURL=(.+)/;
+        const mainSource = compilation.assets[mainSourceFilename].source();
+        const sourceMappingMatch = new RegExp(sourceMappingRegex, 'g').exec(
+          mainSource
+        );
+
+        // Concatenate all chunks and it's source maps. Chunks source needs to have source mapping URL
```
Comment (thymikee) [inline packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts]: Typo: it's -> its

File: packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts [hunk 3]
```diff
@@ -13,26 +15,116 @@ function asyncEval(url) {
...
+      }
+
       (compilation.mainTemplate as any).hooks.requireEnsure.tap(
         'ReactNativeEnvPlugin',
         (source: string) => {
           // The is no `importScripts` in react-native. Replace it with Promise based
           // fetch + eval and return the promise so the webpack module system and bootstrapping
           // logic is not broken.
+          // When creating static bundle, async chunks will be concatenated into the bundle,
+          // so by the time they are required, they should already by loaded into the module system.
```
Comment (thymikee) [inline packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts]: By -> be

File: packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts [hunk 4]
```diff
@@ -13,26 +15,116 @@ function asyncEval(url) {
...
           return source.replace(
             /importScripts\((.+)\)/gm,
-            'return asyncEval(__webpack_require__.p + $1)'
+            this.bundle
+              ? 'throw new Error("async chunks were not concatenated into the bundle")'
```
Comment (thymikee) [inline packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts]: Maybe some hint on what to do in this case, even if it's creating an issue?

File: packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts [hunk 5]
```diff
@@ -13,26 +15,116 @@ function asyncEval(url) {
...
+
+        // Add single source mapping url pointing to file with concatenated source maps.
+        if (sourceMappingMatch) {
+          concat.add(null, sourceMappingMatch[0]);
+        }
+
+        // Remove non-main assets
+        Object.keys(compilation.assets)
+          .filter(item => ![mainSourceFilename, mainMap].includes(item))
+          .forEach(item => {
```
Comment (thymikee) [inline packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts]: Not sure how big this array may get, but if it's 1000+ items it's probably worth to have only one forEach and move filtering inside

Comment (zamotany) [reply packages/haul-core/src/webpack/plugins/ReactNativeEnvPlugin.ts]: Each `import()` creates new asset + main bundle and main bundle maps. Not sure if someone will even use so many `import()` to go beyond few hundreds.

**Pull request merged**